### PR TITLE
Improve cli config file generation

### LINKED ISSUE #1240: Use service discovery for CLI configuration
#### Summary

Add endpoint in The Things Stack for serving the `ttn-lw-cli` config file

#### Why do we need this?

To download the concerning configuration file, most importantly including the host names of the gRPC services and potentially CA to use

#### What is already there? What do you see now?

Manual configuration

#### What is missing? What do you want to see?

A public endpoint to download the `ttn-lw-cli` configuration.

Bonus would be;
- Add a command in the CLI that takes the host name and downloads the `yml` file
- By default saves it to `ttn-lw-cli.yml` in the working dir (confirm overwrite file etc). We can have flags also like `--local` or `--global` to specify where the file should go
- Uses `ttn-lw-cli.yml` from the working dir (before checking `$HOME/.ttn-lw-cli.yml`) if present

That way, users can simply:

```
$ ttn-lw-cli use join.thethings.industries --local
$ ttn-lw-cli login
```

#### How do you propose to implement this?

For serving the configuration, simply a public endpoint to `GET` the configuration. I guess that YAML is all we need, but we may consider other formats through `Accept`

#### Can you do this yourself and submit a Pull Request?

Can review

### PR DESCRIPTION
**Pull request by neoaggelos**
#### Summary

Closes #1240

#### Changes

- Use service discovery for the gRPC service ports
- Add `--grpc-port` and `--oauth-server-address` options in use command.

#### Notes for Reviewers

...

#### Checklist

- [X] Scope: The referenced issue is addressed, there are no unrelated changes.
- [X] Compatibility: The changes are backwards compatible with existing API, database and configuration, according to the stability commitments in `README.md`.
- [X] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [X] Documentation: Relevant documentation is added or updated.
- [X] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [X] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: cmd/ttn-lw-cli/commands/config.go [hunk 1]
```diff
@@ -20,11 +20,21 @@ import (
...
+	"go.thethings.network/lorawan-stack/pkg/rpcmiddleware/discover"
 )
 
+func httpAddress(host string, tls bool) string {
+	if tls {
+		return fmt.Sprintf("https://%s:8885", host)
+	}
+	return fmt.Sprintf("http://%s:1885", host)
+}
+
```
Comment (neoaggelos) [inline cmd/ttn-lw-cli/commands/config.go:32]: This should probably be extracted in a separate package, just like `discover.WithDefaultPort`

Comment (htdvisser) [reply cmd/ttn-lw-cli/commands/config.go:32]: Why don't we just use the same logic for HTTP as we do with GRPC? So actually using `discover.WithDefaultPort`?

File: CHANGELOG.md [hunk 2]
```diff
@@ -9,6 +9,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- You can specify server gRPC port and OAuth server address when generating a CLI config file with `ttn-lw-cli use` (see `--grpc-port` and `--oauth-server-address` options).
```
Comment (johanstokking) [inline CHANGELOG.md:12]: Changelog doesn't talk to the user

File: cmd/ttn-lw-cli/commands/config.go (hunk 1 above)
Comment (johanstokking) [reply cmd/ttn-lw-cli/commands/config.go:32]: This should definitely be in the discovery package

File: pkg/rpcmiddleware/discover/discover.go [hunk 3]
```diff
@@ -28,7 +28,8 @@ import (
 
 var errAddress = errors.DefineInvalidArgument("address", "invalid address")
 
-func defaultPort(target string, port int) (string, error) {
+// WithDefaultPort appends port if target does not already have one.
+func WithDefaultPort(target string, port int) (string, error) {
```
Comment (johanstokking) [inline pkg/rpcmiddleware/discover/discover.go:32]: `With` is reserved for options or funcs that change state

File: cmd/ttn-lw-cli/commands/config.go (hunk 1 above)
Comment (neoaggelos) [reply cmd/ttn-lw-cli/commands/config.go]: For HTTP, we also need the protocol (http/https). Added a `discover.DefaultURL` and using that.

File: CHANGELOG.md (hunk 2 above)
Comment (neoaggelos) [reply CHANGELOG.md]: Updated text

File: pkg/rpcmiddleware/discover/discover.go (hunk 3 above)
Comment (neoaggelos) [reply pkg/rpcmiddleware/discover/discover.go]: Renamed to the original `DefaultPort`

File: pkg/rpcmiddleware/discover/discover.go [hunk 4]
```diff
@@ -49,15 +50,37 @@ func defaultPort(target string, port int) (string, error) {
...
+var DefaultPorts = map[bool]int{
 	false: 1884,
 	true:  8884,
 }
 
+// DefaultHTTPPorts is a map of the default HTTP ports, with/without TLS.
+var DefaultHTTPPorts = map[bool]int{
+	false: 1885,
+	true:  8885,
+}
```
Comment (htdvisser) [inline pkg/rpcmiddleware/discover/discover.go:63]: Not completely sure about these defaults. On local deployments we indeed use 1885/8885, but for online deployments all our guides use 80/443.

Comment (johanstokking) [reply pkg/rpcmiddleware/discover/discover.go:63]: Right. This should indeed be 80/443.

File: pkg/rpcmiddleware/discover/discover.go [hunk 5]
```diff
@@ -49,15 +50,37 @@ func defaultPort(target string, port int) (string, error) {
...
+
+// HTTPProto is a map of the HTTP protocols, with/without TLS.
+var HTTPProto = map[bool]string{
```
Comment (johanstokking) [inline pkg/rpcmiddleware/discover/discover.go:66]: This is generally called a scheme, not a proto(col)

**Comment by neoaggelos** [comment]
> How is this tested?

The discovery package is tested. There are no automated tests for most of the CLI functionality currently. It would be a good thing to have.

**Comment by johanstokking** [comment]
Sure I see the unit tests. But did you test this manually, locally?

**Comment by neoaggelos** [comment]
Just for reference, the changes have been tested locally:

- `grpc-port` can override the default port numbers (8884 or 1884)
- `oauth-server-address` can override the default oauth server address
- `insecure` is checked in order to decide default ports (1xxx vs 8xxx) and http scheme

**Pull request merged**